common/tools/dev-tool/src/framework/check.ts

```diff
@@ -0,0 +1,317 @@
+import path from "node:path";
+import { format } from "../util/prettier";
+import { ProjectInfo } from "../util/resolveProject";
+import { run } from "../util/run";
+import fs from "node:fs/promises";
+
+export type CheckTags = "release" | "ci" | "local";
+export type CheckSeverity = "warning" | "error";
+
+export type CheckEnableFunction = (project: ProjectInfo) => Promise<boolean>;
+
+export interface CheckOptions {
+  /**
+   * Name of the check
+   *
+   * If unspecified a name will be inferred from the export
+   */
+  name?: string;
+
+  /**
+   * Optional extended description of the check
+   */
+  description?: string;
+
+  /**
+   * True if the check can attempt to fix itself
+   */
+  hasFix?: boolean;
+
+  /**
+   * Tags to identify when the check should be run
+   *
+   * Can have the following values:
+   * - "ci": check is intended to be run in CI pipeline
+   * - "release": check is intended to be run before releasing the package
+   * - "local": check is intended to be run locally before pushing commit
+   *
+   * If `tags` is set to "*" then the check will be treated as if it has _every_ tag and will be always run.
+   *
+   * If no tags are specified the check will only be run if no tag is specified
+   */
+  tags?: CheckTags[] | "*";
+
+  /**
+   * Severity of the check. Defaults to "error".
+   */
+  severity?: CheckSeverity;
+
+  /**
+   * Optional function to determine whether this check applies to the given project
+   * If this function returns true for the project, the check will be run.
+   * If the function returns false, the check will be skipped for the project.
+   *
+   * If left undefined, the check will be enabled for all projects.
+   */
+  enable?: CheckEnableFunction;
+}
+
+export interface CheckContext {
+  /**
+   * Whether to attempt to fix
+   */
+  fix: boolean;
+
+  /**
+   * Information about the project
+   */
+  project: ProjectInfo;
+
+  /**
+   * Whether to output verbosely
+   */
+  verbose: boolean;
+}
+
+export type CheckFunction = (ctx: CheckContext) => void | Promise<void>;
+
+export interface Check extends CheckOptions {
+  /**
+   * Run the check in the given context
+   */
+  check: CheckFunction;
+}
+
+/**
+ * Error indicating that a check has failed
+ */
+export class CheckFailedError extends Error {
+  constructor(
+    message: string,
+    public detail?: string,
+  ) {
+    super(message);
+    this.name = "CheckFailedError";
+  }
+}
+
+export function isCheckFailedError(e: unknown): e is CheckFailedError {
+  return (e as any)?.name === "CheckFailedError";
```

Comment:
nitpick: maybe avoid casting to `any`
```suggestion
  return e instanceof Error && e.name === "CheckFailedError";
```